include/nbl/asset/ICPUBuffer.h

```diff
@@ -90,93 +108,30 @@ class ICPUBuffer : public asset::IBuffer, public IPreHashed
-template<
-    typename Allocator = _NBL_DEFAULT_ALLOCATOR_METATYPE<uint8_t>,
-    bool = std::is_same<Allocator, core::null_allocator<typename Allocator::value_type> >::value
->
-class CCustomAllocatorCPUBuffer;
-
-using CDummyCPUBuffer = CCustomAllocatorCPUBuffer<core::null_allocator<uint8_t>, true>;
-
-//! Specialization of ICPUBuffer capable of taking custom allocators
-/*
-    Take a look that with this usage you have to specify custom alloctor
-    passing an object type for allocation and a pointer to allocated
-    data for it's storage by ICPUBuffer.
-
-        So the need for the class existence is for common following tricks - among others creating an
-        \bICPUBuffer\b over an already existing \bvoid*\b array without any \imemcpy\i or \itaking over the memory ownership\i.
-        You can use it with a \bnull_allocator\b that adopts memory (it is a bit counter intuitive because \badopt = take\b ownership,
-        but a \inull allocator\i doesn't do anything, even free the memory, so you're all good).
-    */
-
-template<typename Allocator>
-class CCustomAllocatorCPUBuffer<Allocator,true> : public ICPUBuffer
-{
-        static_assert(sizeof(typename Allocator::value_type) == 1u, "Allocator::value_type must be of size 1");
-    protected:
-        Allocator m_allocator;
-
-        virtual ~CCustomAllocatorCPUBuffer() final
-        {
-            freeData();
-        }
-        inline void freeData() override
-        {
-            if (ICPUBuffer::data)
-                m_allocator.deallocate(reinterpret_cast<typename Allocator::pointer>(ICPUBuffer::data), ICPUBuffer::m_creationParams.size);
-            ICPUBuffer::data = nullptr; // so that ICPUBuffer won't try deallocating
-        }
-
-    public:
-        CCustomAllocatorCPUBuffer(size_t sizeInBytes, void* dat, core::adopt_memory_t, Allocator&& alctr = Allocator()) : ICPUBuffer(sizeInBytes,dat), m_allocator(std::move(alctr))
-        {
-        }
-};
-
-template<typename Allocator>
-class CCustomAllocatorCPUBuffer<Allocator, false> : public CCustomAllocatorCPUBuffer<Allocator, true>
-{
-        using Base = CCustomAllocatorCPUBuffer<Allocator, true>;
-
-    protected:
-        virtual ~CCustomAllocatorCPUBuffer() = default;
-        inline void freeData() override {}
-
-    public:
-        using Base::Base;
-
-        // TODO: remove, alloc can fail, should be a static create method instead!
-        CCustomAllocatorCPUBuffer(size_t sizeInBytes, const void* dat, Allocator&& alctr = Allocator()) : Base(sizeInBytes, alctr.allocate(sizeInBytes), core::adopt_memory, std::move(alctr))
-        {
-            memcpy(Base::data,dat,sizeInBytes);
-        }
+protected:
+    inline IAsset* getDependant_impl(const size_t ix) override
+    {
+        return nullptr;
+    }
+
+    inline void discardContent_impl() override
+    {
+        return freeData();
+    }
+
+    // REMEMBER TO CALL FROM DTOR!
+    virtual inline void freeData()
+    {
+        if (!m_adopt_memory && m_data)

```

Comment:
nooooooooooooooooooooooooooooooooooooo... we **never** checked if the memory was adopted to decide whether to call the deallocation method, WE ALWAYS called it

if  you want to skip deallocation, then use the null memory resource & adopt memory (if you don't adopt it will throw std::bad_alloc during allocation as intended)